lib/cpp/src/thrift/protocol/TProtocol.h

```diff
@@ -659,6 +659,11 @@ class TProtocolFactory {
   virtual ~TProtocolFactory() {}
 
   virtual boost::shared_ptr<TProtocol> getProtocol(boost::shared_ptr<TTransport> trans) = 0;
+  virtual boost::shared_ptr<TProtocol> getProtocol(boost::shared_ptr<TTransport> inTrans,
```

Comment:
THeader detects the client transport type - unframed, framed, header - and responds in kind.  To make this work, the input/output transport need to be the *same* THeader instance.  

The transports don't necessarily need to be the same though (and in the case of TNonblocking, need to be different due to the way the MemoryBuffers work).  So THeader here takes in two transports, but uses a single protocol.

Actually in the fbthrift code this wasn't added to the interface, but just special cased in TNonblockingServer with a dynamic_cast to TheaderProtocol, which we could do instead